### DISCUSSION TITLE
Calculating overlap integrals

### DISCUSSION THREAD

**Discussion by HelgeGehring** [kinnala/scikit-fem#961] [Q&A] [answered]
How would I calculate an integral which depends on two fields which I found as eigenvectors?
The Basis which I use for $E_i$ and $H_i$ is `ElementTriN0() * ElementTriP1()`.

The integral looks like:
![image](https://user-images.githubusercontent.com/42973196/201181539-b6229b9f-8843-45c6-9825-8a6f35d5347b.png)
i.e. a cross-product of which only the z-component is used (x would be a unit vector in z-direction)

If I had only one field in the integral, I'd take a BilinearForm and multiply the assembled matrix by the eigenvector and then sum it up. 
But for the case with 2 fields I don't really know how to do it. Could you give me here a hint?

At the moment I'm thinking about projecting the xy component on a basis with `ElementVector(ElementTriP0())` and then putting those component-wise in a functional. But that seems somehow messy :/

**Comment by kinnala** [comment, marked as answer]
You can pass multiple fields to functional:
```python
# todo: def cross
@Functional
def func(w):
    E_i = w['E_i']
    H_j = w['H_j']
    return ...

func.assemble(E_i=basis.interpolate(E_i), H_j=basis.interpolate(H_j))
```
Do you mean that E_i is ElementTriN0() and H_j is ElementTriP1() or they are both ElementComposite(ElementTriN0(), ElementTriP1())?

**Reply by HelgeGehring** [reply]
Both are ElementComposite,
E and H are in principal (E_x,E_y,E_z), where E_x,E_y are represented by ElementTriN0 and E_z is represented by ElementTriP1

**Reply by HelgeGehring** [reply]
the i/j you can ignore here, from the calculation I get a lot of modes and I want to see the overlaps between different modes
(of course, if I calculate overlaps between modes from the same geometry they are either 1 or 0, but if I have a waveguide abruptly change, I can like this calculate the transmission to the modes of the new waveguide)

**Reply by kinnala** [reply]
inside the integral you have both H_i and H_j, and same for E. So you need to pass four fields?

**Reply by HelgeGehring** [reply]
ups, true four field, from $E_i$ and $E_j$ i can calculate $H_i$ and $H_j$

**Reply by HelgeGehring** [reply]
But thanks! Now I see how I can calculate the integrals, I forgot to add basis.interpolate and only got vectors inside the form, which confused me

**Comment by HelgeGehring** [comment]
Would it be possible to calculate such integrals with vectors from different meshes?
I.e. if I calculate one eigenvector (E_i/H_i) on one mesh and (E_j/H_j) on another mesh - how would I then calculate such a functional?

I would need for one eigenvector to calculate the values at the quadrature points of the other basis and then I could use them in the same functional, right?
So there would just need to be a way interpolate on quadrature points of another basis?

**Reply by kinnala** [reply]
If that's enough, yes, it is possible using Basis.interpolator. However interpolating at the quadrature points of another mesh will introduce integration error and sometimes this error will dominate. This is why I'm working on skfem.supermeshing and the related kinnala/supermeshing package. They allow exact integration between two meshes. Unfortunately these are not ready yet and moreover kinnala/supermeshing will require Fortran code to do it efficiently.

**Reply by HelgeGehring** [reply]
Thanks! Then we'll directly create a mesh which supports both geometries :)
Nice to hear that there'll be a solution in skfem soon :) Even through it's Fortran based :/

**Reply by kinnala** [reply]
Technically possible to do supermeshing with scipy rtree and some vectorized splitting of all triangles at one go. However, I don't have the energy right now to think about it.

**Reply by HelgeGehring** [reply]
Nice to hear that there's also a way to a scipy solution :)

**Reply by HelgeGehring** [reply]
I'm currently trying to use Basis.interpolator, I'm getting `NotImplementedError: TODO: fix tind support in ElementHcurl` :/ Would we need to implement this or is there another way for the interpolation?

**Reply by kinnala** [reply]
Alright, before implementing subset assembly you need to project on some ElementH1 and use interpolator there. This will also affect boundary assembly.

**Reply by HelgeGehring** [reply]
Hmm, I don't completely understand your answer, sorry :/
At the moment I'm projecting on a ElementP1 and then I'm using the Interpolator. It somehow works, but I'd guess it increases the error as I'm interpolating from the projected values, right? It also makes the code more complicated :(

**Reply by kinnala** [reply]
Yes, but projection is exact if you pick the target space that has the correct polynomial degree so there shouldn't be additional error introduced.

**Reply by kinnala** [reply]
For example, if you want to project the derivative of ElementTriP2 you would use ElementDG(ElementTriP1) because that is the correct space where the weak derivative exists. In such case the projection should exactly represent the pointwise derivative of the discrete firld.

**Reply by HelgeGehring** [reply]
Thanks!
Right means N1->P1, N2->P2?
Would it work directly if the `tind` is implemented or is there a reason that the projection is necessary?

**Reply by kinnala** [reply]
I believe the components of ElementTriN2 are in ElementDG(ElementTriP2).

**Reply by kinnala** [reply]
Sure it is also possible to implement the slicing part in class ElementHcurl. The main idea is that the routine should be able to map also a subset of elements from reference element to global element. Normally it maps all elements at once which is usually slightly faster because there is no numpy slicing going on. You can check ElementH1 and ElementHdiv how it's done there.

However, the next issue you will run into is #973. Perhaps there are other issues forthcoming because no vector interpolation has been done before.
